# COMP 4462 Data Visualization Tutorial 5: Pandas and Visualization

In this tutorial, we will learn more about Pandas and Data Visualization with Python.

You can view this notebook on Google Colab ([https://bit.ly/vis-t05-nb](https://bit.ly/vis-t05-nb)).

All the materials of this tutorial are hosted on [this GitHub repository](https://github.com/leoyuholo/learning-vis-tools/tree/master/tutorial05).

## Dataset

### Download dataset

We will explore the [Spotify"s Worldwide Daily Song Ranking](https://www.kaggle.com/edumucelli/spotifys-worldwide-daily-song-ranking) prepared by Eduardo on Kaggle.

Since the dataset is not tiny enough for the whole class to upload simultaneously, it is shared on Google Drive (https://drive.google.com/open?id=1ki4HKtNKpj-Ewj5RGToXpMrzgXh4ZIbz) and we will download it directly into the Google Colab workspace. The zipped file is about 43MB and the unzipped file contains more than 3 million records and size over 350MB.

Any commands that starts with `!` are shell commands, they are not Python code. Instead, you can run commands like `ls`, `mkdir`, etc. in Jupyter notebook. In the following case, we use `pip`, a command to install Python package, then use the installed package `gdown` to download the file shared on Google Drive. Finally, we unzip the file to obtain the dataset we will use throughout this tutorial.

In [8]:
# Download dataset, it will take about 1 min

!pip -q install gdown
!gdown https://drive.google.com/uc?id=1ki4HKtNKpj-Ewj5RGToXpMrzgXh4ZIbz
!unzip spotifys-worldwide-daily-song-ranking.zip

## Explore the dataset with Pandas

In [9]:
# Import pandas

import pandas as pd

In [10]:
# Since the dataset is about 350MB and contains over 3 million records, it will 
# take some time to load the dataset

df_daily_ranking = pd.read_csv("data.csv")

In [11]:
# Show the first 5 rows of the data

df_daily_ranking.head()

Position                  Track Name         Artist  Streams  \
0         1  Reggaetón Lento (Bailemos)           CNCO    19272   
1         2                    Chantaje        Shakira    19270   
2         3   Otra Vez (feat. J Balvin)  Zion & Lennox    15761   
3         4                Vente Pa' Ca   Ricky Martin    14954   
4         5                      Safari       J Balvin    14269   

                                                 URL        Date Region  
0  https://open.spotify.com/track/3AEZUABDXNtecAO...  2017-01-01     ec  
1  https://open.spotify.com/track/6mICuAdrwEjh6Y6...  2017-01-01     ec  
2  https://open.spotify.com/track/3QwBODjSEzelZyV...  2017-01-01     ec  
3  https://open.spotify.com/track/7DM4BPaS7uofFul...  2017-01-01     ec  
4  https://open.spotify.com/track/6rQSrBHf7HlZjtc...  2017-01-01     ec

In [12]:
# Compute the statistical summary of all the columns

df_daily_ranking.describe(include="all")

Position    Track Name      Artist       Streams  \
count   3.441197e+06       3440540     3440540  3.441197e+06   
unique           NaN         18597        6628           NaN   
top              NaN  Shape of You  Ed Sheeran           NaN   
freq             NaN         19365      127064           NaN   
mean    9.464399e+01           NaN         NaN  5.189176e+04   
std     5.739567e+01           NaN         NaN  2.018035e+05   
min     1.000000e+00           NaN         NaN  1.001000e+03   
25%     4.500000e+01           NaN         NaN  3.322000e+03   
50%     9.200000e+01           NaN         NaN  9.227000e+03   
75%     1.430000e+02           NaN         NaN  2.965800e+04   
max     2.000000e+02           NaN         NaN  1.138152e+07   

                                                      URL        Date   Region  
count                                             3441189     3441197  3441197  
unique                                              21746         371       54  
top     https://open.spotify.com/track/7qiZfU4dY1lWllz...  2017-12-24       ec  
freq                                                19365       10073    74200  
mean                                                  NaN         NaN      NaN  
std                                                   NaN         NaN      NaN  
min                                                   NaN         NaN      NaN  
25%                                                   NaN         NaN      NaN  
50%                                                   NaN         NaN      NaN  
75%                                                   NaN         NaN      NaN  
max                                                   NaN         NaN      NaN

## Clean up

### See if any rows contain null

In [13]:
# Show 5 rows that contain null in any columns
# We will revisit the this selector syntax later

df_daily_ranking[df_daily_ranking.isnull().any(axis=1)].head()

Position Track Name Artist  Streams  \
39428        29        NaN    NaN     7362   
39456        57        NaN    NaN     4426   
39463        64        NaN    NaN     4069   
39490        91        NaN    NaN     3238   
39631        32        NaN    NaN     7409   

                                                     URL        Date Region  
39428  https://open.spotify.com/track/3RXkboS74UYzN14...  2017-07-20     ec  
39456  https://open.spotify.com/track/4JAyIDXOqNM6qHu...  2017-07-20     ec  
39463  https://open.spotify.com/track/3bVbQvGVIe4n24A...  2017-07-20     ec  
39490  https://open.spotify.com/track/3eFJqPe8VUYrABb...  2017-07-20     ec  
39631  https://open.spotify.com/track/3RXkboS74UYzN14...  2017-07-21     ec

In [14]:
# Counting how many records have null values, for each column

df_daily_ranking.isnull().sum()

Position        0
Track Name    657
Artist        657
Streams         0
URL             8
Date            0
Region          0
dtype: int64

### Drop the null rows

In [15]:
# Drop all the rows that contain any null values
# Inplace avoid copying the whole dataframe, computationally faster

df_daily_ranking.dropna(inplace=True)

### Separate "global" from the rest

In the dataset, there is a region "global", which is the sum of all the streams all over the world. It overlaps with other regions, treating it as equally as a region will cause "double count", so we separate it out.

In [16]:
df_daily_ranking_global = df_daily_ranking[df_daily_ranking['Region'] == 'global']
df_daily_ranking = df_daily_ranking[df_daily_ranking['Region'] != 'global']

## Aggregate

### Top artists of the year

In [17]:
# Count how many unique artists in the dataset

df_daily_ranking["Artist"].nunique()

6628

The following command has multiple components, `groupby`, `sum`, `sort_values`, column selection and `head`, we look into it one by one:

- `groupby`: This is like splitting the whole dataset by the "Artist" of each row, resulting in 6628 groups.
- `sum`: As there are multiple rows in each group of 6628 groups, this command sums up the values of each group. In our case, it sums up the "Streams" column of each row in each group.
- `sort_values`: After summing up, each group becomes one row of record, that is 6628 rows now. This command sort these 6628 rows by their value in "Streams" in descending order.
- select the "Streams" column: This is like ignoring other columns and only show us the "Streams" column alongside with the index column (which is the column used in the `groupby` command).
- `head`: Show the first 5 rows

In [18]:
# Show the top 5 artists of the year

df_daily_ranking.groupby("Artist").sum().sort_values("Streams", ascending=False)["Streams"].head()

Artist
Ed Sheeran          4353885528
Drake               2285102445
The Chainsmokers    2081716050
Post Malone         1865412162
Luis Fonsi          1760377876
Name: Streams, dtype: int64

### Top songs of the year

In [19]:
################################################################################
# TODO:                                                                        #
# Try to show the top 5 songs of the year!                                     #
################################################################################

df_daily_ranking.groupby("Track Name").sum().sort_values("Streams", ascending=False)["Streams"].head()

################################################################################
#                                 END OF YOUR CODE                             #
################################################################################

Track Name
Shape of You                          1503176575
Despacito - Remix                      910422437
Despacito (Featuring Daddy Yankee)     717960928
Something Just Like This               693853930
Unforgettable                          688520127
Name: Streams, dtype: int64

## Filter

### Top song of the year in US

The square brackets of pandas have multiple usage, it is sometimes even confusing. For now, we look into using it as column selection and filtering.

If the value inside the brackets is a string, it will select the column, just like `df_daily_ranking["Region"]` will return a column.

If the value inside the brackets is an array of boolean values, it will return only the `True` rows.

Let"s see an example!

In [20]:
# Ignore the code below at the moment, all we care is "df_sample" now contains 
# 10 rows, and 5 are in Region "us" and 5 in Region "hk"

df_sample = pd.concat([df_daily_ranking[df_daily_ranking["Region"] == "us"].head(), df_daily_ranking[df_daily_ranking["Region"] == "hk"].head()])
df_sample

Position                                         Track Name  \
771622          1                Bad and Boujee (feat. Lil Uzi Vert)   
771623          2                                          Fake Love   
771624          3                                            Starboy   
771625          4                                             Closer   
771626          5                                      Black Beatles   
3366997         1                                             Closer   
3366998         2                                            Starboy   
3366999         3                                    Let Me Love You   
3367000         4  I Don’t Wanna Live Forever (Fifty Shades Darke...   
3367001         5                                   Don't Wanna Know   

                   Artist  Streams  \
771622              Migos  1371493   
771623              Drake  1180074   
771624         The Weeknd  1064351   
771625   The Chainsmokers  1010492   
771626       Rae Sremmurd   874289   
3366997  The Chainsmokers    12831   
3366998        The Weeknd     9626   
3366999          DJ Snake     9499   
3367000              ZAYN     8210   
3367001          Maroon 5     7606   

                                                       URL        Date Region  
771622   https://open.spotify.com/track/4Km5HrUvYTaSUfi...  2017-01-01     us  
771623   https://open.spotify.com/track/343YBumqHu19cGo...  2017-01-01     us  
771624   https://open.spotify.com/track/5aAx2yezTd8zXrk...  2017-01-01     us  
771625   https://open.spotify.com/track/7BKLCZ1jbUBVqRi...  2017-01-01     us  
771626   https://open.spotify.com/track/6fujklziTHa8uoM...  2017-01-01     us  
3366997  https://open.spotify.com/track/7BKLCZ1jbUBVqRi...  2017-01-01     hk  
3366998  https://open.spotify.com/track/5aAx2yezTd8zXrk...  2017-01-01     hk  
3366999  https://open.spotify.com/track/4pdPtRcBmOSQDlJ...  2017-01-01     hk  
3367000  https://open.spotify.com/track/3NdDpSvN911VPGi...  2017-01-01     hk  
3367001  https://open.spotify.com/track/5MFzQMkrl1FOOng...  2017-01-01     hk

In [21]:
# To see how filter works, we now see how the "==" operator works

df_sample["Region"] == "us"

771622      True
771623      True
771624      True
771625      True
771626      True
3366997    False
3366998    False
3366999    False
3367000    False
3367001    False
Name: Region, dtype: bool

While the left column is the row id, the right column is the truth value of whether the row has "Region" equals "us". By passing this array into the dataframe, it will filter out only the rows marked as `True`, that is the first 5 rows.

In [22]:
# Only keep rows with "Region" equals "us"

df_sample[df_sample["Region"] == "us"]

Position                           Track Name            Artist  \
771622         1  Bad and Boujee (feat. Lil Uzi Vert)             Migos   
771623         2                            Fake Love             Drake   
771624         3                              Starboy        The Weeknd   
771625         4                               Closer  The Chainsmokers   
771626         5                        Black Beatles      Rae Sremmurd   

        Streams                                                URL  \
771622  1371493  https://open.spotify.com/track/4Km5HrUvYTaSUfi...   
771623  1180074  https://open.spotify.com/track/343YBumqHu19cGo...   
771624  1064351  https://open.spotify.com/track/5aAx2yezTd8zXrk...   
771625  1010492  https://open.spotify.com/track/7BKLCZ1jbUBVqRi...   
771626   874289  https://open.spotify.com/track/6fujklziTHa8uoM...   

              Date Region  
771622  2017-01-01     us  
771623  2017-01-01     us  
771624  2017-01-01     us  
771625  2017-01-01     us  
771626  2017-01-01     us

In [23]:
# By applying the same logic, we filter the whole dataset

df_daily_ranking_in_us = df_daily_ranking[df_daily_ranking["Region"] == "us"]

# Then, do the aggregation and find out the top songs in US

df_daily_ranking_in_us.groupby("Artist").sum().sort_values("Streams", ascending=False)["Streams"].head()

Artist
Drake             1250864578
Kendrick Lamar    1163890899
Post Malone        988811897
Lil Uzi Vert       773675118
Ed Sheeran         723507889
Name: Streams, dtype: int64

### Top song of the year in Hong Kong

In [24]:
################################################################################
# TODO:                                                                        #
# Try to show the top 5 songs of the year in Hong Kong!                        #
################################################################################

df_daily_ranking_in_hk = df_daily_ranking[df_daily_ranking["Region"] == "hk"]
df_daily_ranking_in_hk.groupby("Track Name").sum().sort_values("Streams", ascending=False)["Streams"].head()

################################################################################
#                                 END OF YOUR CODE                             #
################################################################################

Track Name
Shape of You                5042866
Something Just Like This    3867676
Despacito - Remix           3104779
Closer                      2624849
Attention                   2461646
Name: Streams, dtype: int64

## Read dataset in JSON format

In order to find out who are the top artists or the top songs in Asia / North America / Europe, we need to link up the country code to continents. To do so, we need an extra dataset. You can download the `country.json` from the [tutorial material repository on GitHub](https://github.com/leoyuholo/learning-vis-tools/tree/master/tutorial04/lab3). Thanks to the authors of the GitHub repository [annexare/Countries](https://github.com/annexare/Countries).

If you are using Google Colab, you should be able to see an arrow on the left edge of the window, it hides the panel. Open it and switch to the "Files" tab, you can see the files in the current directory. Click upload and upload the dataset you have downloaded.

In [25]:
# Read JSON file into dataframe

df_countries = pd.read_json("countries.json")
df_countries

AD                             AE            AF  \
name                Andorra           United Arab Emirates   Afghanistan   
native              Andorra  دولة الإمارات العربية المتحدة     افغانستان   
phone                   376                            971            93   
continent                EU                             AS            AS   
capital    Andorra la Vella                      Abu Dhabi         Kabul   
currency                EUR                            AED           AFN   
languages              [ca]                           [ar]  [ps, uz, tk]   

                            AG          AI         AL        AM      AO  \
name       Antigua and Barbuda    Anguilla    Albania   Armenia  Angola   
native     Antigua and Barbuda    Anguilla  Shqipëria  Հայաստան  Angola   
phone                     1268        1264        355       374     244   
continent                   NA          NA         EU        AS      AF   
capital           Saint John's  The Valley     Tirana   Yerevan  Luanda   
currency                   XCD         XCD        ALL       AMD     AOA   
languages                 [en]        [en]       [sq]  [hy, ru]    [pt]   

                   AQ            AR  ...        VN            VU  \
name       Antarctica     Argentina  ...   Vietnam       Vanuatu   
native     Antarctica     Argentina  ...  Việt Nam       Vanuatu   
phone             672            54  ...        84           678   
continent          AN            SA  ...        AS            OC   
capital                Buenos Aires  ...     Hanoi     Port Vila   
currency                        ARS  ...       VND           VUV   
languages          []      [es, gn]  ...      [vi]  [bi, en, fr]   

                          WF        WS                   XK       YE  \
name       Wallis and Futuna     Samoa               Kosovo    Yemen   
native      Wallis et Futuna     Samoa  Republika e Kosovës  اليَمَن   
phone                    681       685      377,381,383,386      967   
continent                 OC        OC                   EU       AS   
capital             Mata-Utu      Apia             Pristina   Sana'a   
currency                 XPF       WST                  EUR      YER   
languages               [fr]  [sm, en]             [sq, sr]     [ar]   

                  YT                                        ZA      ZM  \
name         Mayotte                              South Africa  Zambia   
native       Mayotte                              South Africa  Zambia   
phone            262                                        27     260   
continent         AF                                        AF      AF   
capital    Mamoudzou                                  Pretoria  Lusaka   
currency         EUR                                       ZAR     ZMK   
languages       [fr]  [af, en, nr, st, ss, tn, ts, ve, xh, zu]    [en]   

                     ZW  
name           Zimbabwe  
native         Zimbabwe  
phone               263  
continent            AF  
capital          Harare  
currency            ZWL  
languages  [en, sn, nd]  

[7 rows x 250 columns]

In [26]:
# Flip the rows and columns

df_countries.transpose().head()

name                         native phone continent  \
AD               Andorra                        Andorra   376        EU   
AE  United Arab Emirates  دولة الإمارات العربية المتحدة   971        AS   
AF           Afghanistan                      افغانستان    93        AS   
AG   Antigua and Barbuda            Antigua and Barbuda  1268        NA   
AI              Anguilla                       Anguilla  1264        NA   

             capital currency     languages  
AD  Andorra la Vella      EUR          [ca]  
AE         Abu Dhabi      AED          [ar]  
AF             Kabul      AFN  [ps, uz, tk]  
AG      Saint John's      XCD          [en]  
AI        The Valley      XCD          [en]

In [27]:
# The argument "orient" does the exactly same thing

df_countries = pd.read_json("countries.json", orient="index")

## Join

In [28]:
# Since the country code is in lower case, to match up, we need to make them 
# upper case

df_daily_ranking["country"] = df_daily_ranking["Region"].str.upper()

In [29]:
# Merge the two dataframe on "country" column of df_daily_ranking and on the 
# index column of df_country

df_daily_ranking_with_continent = df_daily_ranking.merge(df_countries, how="inner", left_on="country", right_index=True)

In [30]:
# Rename column, just as an example, no specific purpose

df_daily_ranking_with_continent = df_daily_ranking_with_continent.rename(columns={"continent": "Continent"})

### Top artists of the year in North America

In [31]:
# Agggregate just the same as region

df_daily_ranking_in_na = df_daily_ranking_with_continent[df_daily_ranking_with_continent["Continent"] == "NA"]
df_daily_ranking_in_na.groupby("Artist").sum().sort_values("Streams", ascending=False)["Streams"].head(20)

Artist
Drake               1488701744
Kendrick Lamar      1281915448
Post Malone         1125828758
Ed Sheeran          1083913228
Lil Uzi Vert         837881663
Migos                764677867
The Chainsmokers     762650964
Future               627364617
The Weeknd           599949693
Luis Fonsi           582328513
21 Savage            524445998
Khalid               503672138
Kodak Black          482252053
Travis Scott         471269591
J Balvin             465769112
Calvin Harris        459310805
DJ Khaled            456748400
XXXTENTACION         439677187
Imagine Dragons      430717997
Big Sean             428376262
Name: Streams, dtype: int64

### Top song of the year in Asia

In [32]:
################################################################################
# TODO:                                                                        #
# Try to show the top 5 songs of the year in Asia!                             #
################################################################################

df_daily_ranking_in_asia = df_daily_ranking_with_continent[df_daily_ranking_with_continent["Continent"] == "AS"]
df_daily_ranking_in_asia.groupby("Track Name").sum().sort_values("Streams", ascending=False)["Streams"].head()

################################################################################
#                                 END OF YOUR CODE                             #
################################################################################

Track Name
Shape of You                144721255
Something Just Like This     95425270
Despacito - Remix            87475052
Say You Won't Let Go         83336834
That's What I Like           77579196
Name: Streams, dtype: int64

## Pivot Table

In [33]:
# Filter the artists

df_favorite_artist = df_daily_ranking_with_continent[df_daily_ranking_with_continent["Artist"] == "Coldplay"]
df_favorite_artist.head()

Position                         Track Name    Artist  Streams  \
83         84  Hymn For The Weekend - Seeb Remix  Coldplay     2327   
185       186                      The Scientist  Coldplay     1285   
195       196            Adventure Of A Lifetime  Coldplay     1236   
268        69  Hymn For The Weekend - Seeb Remix  Coldplay     2519   
301       102                      The Scientist  Coldplay     1998   

                                                   URL        Date Region  \
83   https://open.spotify.com/track/1OAiWI2oPmglaOi...  2017-01-01     ec   
185  https://open.spotify.com/track/75JFxkI2RXiU7L9...  2017-01-01     ec   
195  https://open.spotify.com/track/69uxyAqqPIsUyTO...  2017-01-01     ec   
268  https://open.spotify.com/track/1OAiWI2oPmglaOi...  2017-01-02     ec   
301  https://open.spotify.com/track/75JFxkI2RXiU7L9...  2017-01-02     ec   

    country     name   native phone Continent capital currency languages  
83       EC  Ecuador  Ecuador   593        SA   Quito      USD      [es]  
185      EC  Ecuador  Ecuador   593        SA   Quito      USD      [es]  
195      EC  Ecuador  Ecuador   593        SA   Quito      USD      [es]  
268      EC  Ecuador  Ecuador   593        SA   Quito      USD      [es]  
301      EC  Ecuador  Ecuador   593        SA   Quito      USD      [es]

In [34]:
# Pivot table, just like spreadsheet, define which attribute goes to rows, which
# to columns, values, and last but not least, how to aggregate them. Aggregation
# can be sum, max, min, mean, etc. In our case, we want to add up the streams 
# counts, so we use "sum"

df_favorite_artist_streams = df_favorite_artist.pivot_table(values="Streams", index="Date", columns="Continent", aggfunc="sum")
df_favorite_artist_streams.head()

Continent         AS        EU        NA       OC        SA
Date                                                       
2017-01-01  134203.0  427319.0  274971.0  26141.0  215578.0
2017-01-02  155686.0  512328.0  318189.0  28136.0  293633.0
2017-01-03  174060.0  525427.0  370151.0  28885.0  318359.0
2017-01-04  197094.0  531836.0  353325.0  32528.0  324242.0
2017-01-05  198938.0  504666.0  345913.0  28291.0  320139.0

In [35]:
################################################################################
# TODO:                                                                        #
# Try to make a pivot table for your favorite song!                            #
################################################################################

df_favorite_song = df_daily_ranking_with_continent[df_daily_ranking_with_continent["Track Name"] == "Havana"]
df_favorite_song_streams = df_favorite_song.pivot_table(values="Streams", index="Date", columns="Continent", aggfunc="sum")
df_favorite_song_streams.head()

################################################################################
#                                 END OF YOUR CODE                             #
################################################################################

Continent        AS        EU        NA  OC        SA
Date                                                 
2017-08-03      NaN    3368.0       NaN NaN  107300.0
2017-08-04  75750.0  265527.0   31688.0 NaN  139815.0
2017-08-05  46171.0  121512.0  183558.0 NaN   95646.0
2017-08-06  43125.0  100622.0  173957.0 NaN   68543.0
2017-08-07  78229.0  141006.0  203847.0 NaN   73237.0

## Altair

In [36]:
import altair as alt

In [37]:
# In case you hit the error message of 5000 rows limitation, you can run the 
# following command to disable the limitation

# alt.data_transformers.enable('default', max_rows=None)

In [38]:
# Same "groupby" trick we used before. The attributes used for "groupby" will 
# becomes index of the dataframe, which are no longer normal columns and cannot 
# be used directly. The method "reset_index" pulls out the attributes used for 
# "groupby" back to normal columns

df_chart = df_favorite_artist.groupby(['Date', 'Continent']).sum()['Streams'].reset_index()
df_chart.head(10)

Date Continent  Streams
0  2017-01-01        AS   134203
1  2017-01-01        EU   427319
2  2017-01-01        NA   274971
3  2017-01-01        OC    26141
4  2017-01-01        SA   215578
5  2017-01-02        AS   155686
6  2017-01-02        EU   512328
7  2017-01-02        NA   318189
8  2017-01-02        OC    28136
9  2017-01-02        SA   293633

Plot dataframe as line chart. The altair API is designed in visualization language. We can apply the knowledge learned in lectures, what we want to encode with each encoding channels. See the documentation for [marks](https://altair-viz.github.io/user_guide/marks.html) and [encodings](https://altair-viz.github.io/user_guide/encoding.html).

In [39]:
# State what we want the X-axis to encode, and what for the Y-axis, finally, 
# what for the color channel
# Data types ("Q": quantitative, "O": ordinal, "N": nominal, "T": 
# temporal) are stated with the delimiter ":" after the column name
# See documentation for more details

alt.Chart(df_chart).mark_line().encode(
    x="Date:T",
    y="Streams:Q",
    color="Continent:N"
)

alt.Chart(...)

### Juxtaposition (side-by-side)

Altair makes it very easy to put multiple charts together, see the [documentation](https://altair-viz.github.io/user_guide/compound_charts.html) for more examples and how to use this powerful feature.

In [40]:
# Filter out the song we want to visualize

df_chart = df_daily_ranking_global[df_daily_ranking_global["Track Name"] == "Something Just Like This"]
df_chart.head()

Position                Track Name            Artist  Streams  \
3127244         2  Something Just Like This  The Chainsmokers  4752225   
3127444         2  Something Just Like This  The Chainsmokers  4460815   
3127644         2  Something Just Like This  The Chainsmokers  3949869   
3127844         2  Something Just Like This  The Chainsmokers  4374275   
3128044         2  Something Just Like This  The Chainsmokers  4517368   

                                                       URL        Date  Region  
3127244  https://open.spotify.com/track/6RUKPb4LETWmmr3...  2017-02-24  global  
3127444  https://open.spotify.com/track/6RUKPb4LETWmmr3...  2017-02-25  global  
3127644  https://open.spotify.com/track/6RUKPb4LETWmmr3...  2017-02-26  global  
3127844  https://open.spotify.com/track/6RUKPb4LETWmmr3...  2017-02-27  global  
3128044  https://open.spotify.com/track/6RUKPb4LETWmmr3...  2017-02-28  global

In [41]:
# We store the return values in variables, and then we can use "|", "&" and "+" 
# to combine them as horizontally, vertically or overlay, see documentation for 
# more details

position_chart = alt.Chart(df_chart).mark_line().encode(
    x='Date:T',
    y=alt.Y('Position:Q', scale=alt.Scale(zero=False, padding=1, domain=[200, 1]))
)

streams_chart = alt.Chart(df_chart).mark_bar().encode(
    x='Date:T',
    y='Streams:Q'
)

position_chart | streams_chart

alt.HConcatChart(...)

In [42]:
################################################################################
# TODO:                                                                        #
# Try to visualize the trend of your favorite song!                            #
################################################################################

df_chart = df_daily_ranking_global[df_daily_ranking_global["Track Name"] == "Havana"]
df_chart.head()

position_chart = alt.Chart(df_chart).mark_line().encode(
    x='Date:T',
    y=alt.Y('Position:Q', scale=alt.Scale(zero=False, padding=1, domain=[200, 1]))
)

################################################################################
#                                 END OF YOUR CODE                             #
################################################################################

### Interactions

Interaction is very useful for exploring datasets. It was time consuming to make charts interactive, thanks to Altair (and the vega-lite behind the scene), it is much easier now. We will go through two examples, but there are more in the [documentation](https://altair-viz.github.io/user_guide/interactions.html), check it out!

In [43]:
# Again, filter, groupby, aggregate, reset_index, apply the tricks we have 
# learnt in this tutorial

df_chart = df_favorite_artist[df_favorite_artist['Continent'] == 'EU'].groupby(['Date', 'Track Name']).sum()['Streams'].reset_index()
df_chart.head()

Date                             Track Name  Streams
0  2017-01-01                Adventure Of A Lifetime    64682
1  2017-01-01              Everglow - Single Version    17622
2  2017-01-01  Everglow - Single Version, Radio Edit    48726
3  2017-01-01                   Hymn For The Weekend    82932
4  2017-01-01      Hymn For The Weekend - Seeb Remix   203571

In [44]:
# Add a scale interaction to allow pan and zoom. Use your scroll wheel to zoom  
# and drag-and-drop to pan. More interaction techniques in documentation

scales = alt.selection_interval(bind='scales')
alt.Chart(df_chart).mark_line().encode(
    x="Date:T",
    y="Streams:Q",
    color="Track Name:N"
).add_selection(
    scales
)

alt.Chart(...)

In [45]:
# Interactively focus on data of interest (gray-out other data). We need to make
# a chart, and a selector ponel. The recommended way is to make two charts, 
# putting them side-by-side, just like the juxtaposition example above. Try to 
# click on the squares in the lengend, it acts like a highlighter.

selection = alt.selection_multi(fields=['Track Name'])
color = alt.condition(selection,
                      alt.Color('Track Name:N', legend=None),
                      alt.value('lightgray'))
chart = alt.Chart(df_chart).mark_line().encode(
    x="Date:T",
    y="Streams:Q",
    color=color,
    tooltip=["Track Name", "Streams", "Date"]
)

legend = alt.Chart(df_chart).mark_square().encode(
    y=alt.Y('Track Name:N', axis=alt.Axis(orient='right')),
    color=color
).add_selection(
    selection
)

chart | legend

alt.HConcatChart(...)